CMakeLists.txt

```diff
@@ -231,51 +231,91 @@ if(BUILD_PYTHON)
   )
 endif()
 
-set(NVFUSER_CODEGEN ${PROJECT_NAME}_codegen)
-add_library(${NVFUSER_CODEGEN} SHARED ${NVFUSER_SRCS})
+# We create both static and shared libraries.
+#
+# Shared libraries are what ships, but a large advantage of static libraries is
+# that symbols are all visible. This allows us to test internal components
+# inside our test or benchmark binaries, even if we do not want said components
+# to be visible to the outside. If we used only shared libraries, then any API
+# we invoked from test binaries would need to be marked as public, even if we
+# did not want to expose it to users.
+#
+# Note technically we create an "OBJECT" library instead of a "STATIC" library.
+# This is just a CMake quirk; an OBJECT library is a better way to implement a
+# "private" (not installed) static library.
+add_library(codegen_internal OBJECT ${NVFUSER_SRCS})
 
 if(NOT MSVC)
-  target_compile_options(${NVFUSER_CODEGEN} PRIVATE
+  target_compile_options(codegen_internal PRIVATE
     -Wall -Wno-unused-function
     -Werror
   )
 endif()
 
-target_compile_definitions(${NVFUSER_CODEGEN} PRIVATE "-DTORCH_CUDA_BUILD_MAIN_LIB")
-target_include_directories(${NVFUSER_CODEGEN} SYSTEM PRIVATE
-  ${CMAKE_SOURCE_DIR}/third_party/flatbuffers/include
+target_compile_definitions(codegen_internal PRIVATE "-DTORCH_CUDA_BUILD_MAIN_LIB")
+target_include_directories(codegen_internal SYSTEM PUBLIC
   ${CMAKE_SOURCE_DIR}/third_party/gloo # TODO: guard this on usage
+  ${CMAKE_SOURCE_DIR}/third_party/flatbuffers/include
+  PRIVATE
   ${CUDA_TOOLKIT_ROOT_DIR}/extras/CUPTI/include
   ${CUDA_INCLUDE_DIRS}
 )
-target_include_directories(${NVFUSER_CODEGEN} PUBLIC
+target_include_directories(codegen_internal PUBLIC
   "$<BUILD_INTERFACE:${NVFUSER_SRCS_DIR}>"
   "$<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/nvfuser>"
 )
-set_property(TARGET ${NVFUSER_CODEGEN} PROPERTY CXX_STANDARD_REQUIRED ON)
-set_property(TARGET ${NVFUSER_CODEGEN} PROPERTY CXX_STANDARD ${NVFUSER_CPP_STANDARD})
+set_target_properties(codegen_internal PROPERTIES
+  C_STANDARD 17
+  CUDA_STANDARD 17 # CUDA 11 does not support C++20
```

Comment:
> maybe a global variable for C/CUDA_STANDARD

it's good that you flagged this because IIRC we recently *did* create a variable for this. It's probably I missed this while merging. I will fix that now.

> if just use a global flag with set_property would work easier.

CMake's global flags are broken and really should never be used. the probably is that they apply to include()d subdirs and project()s that aren't necessarily expecting something to come in and change things on them. https://cliutils.gitlab.io/modern-cmake/chapters/intro/dodonot.html

Reply:
>> maybe a global variable for C/CUDA_STANDARD

> it's good that you flagged this because IIRC we recently did create a variable for this. It's probably I missed this while merging. I will fix that now.

okay, 😌 , I *did* properly merge in the use of `NVFUSER_CPP_STANDARD`; this is specifically about CUDA_STANDARD, which I missed upon first reading your comment.

A variable is still a good idea for these settings too, but because this patch has so many conflicts && is hard to get in, I'll wait for a follow-up PR unless CI is unhappy or similar.

Reply:
Thanks for educating me on CMake's sharp edges regarding global flags. (But really which part of CMake is smooth?!)

Yeah let's merge it first and polish afterwards. I have another PR merged right before yours and I figure my api SHOULD have been exposed as well and I likely have my code base broken.

Reply:
https://github.com/NVIDIA/Fuser/pull/1802